Modules/_decimal/_decimal.c

```diff
@@ -3675,34 +3673,54 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
-#if PYLONG_BITS_IN_DIGIT == 30
-    n = mpd_qexport_u32(&ob_digit, 0, PyLong_BASE, x, &status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    n = mpd_qexport_u16(&ob_digit, 0, PyLong_BASE, x, &status);
-#else
-    #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
+    int64_t val = mpd_qget_i64(x, &status);
 
-    if (n == SIZE_MAX) {
-        PyErr_NoMemory();
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+    assert(!mpd_iszero(x));
+
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+
+    assert(layout->bits_per_digit < 32);
+    assert(layout->digits_order == -1);
+    assert(layout->digit_endianness == (PY_LITTLE_ENDIAN ? -1 : 1));
+    assert(layout->digit_size == 2 || layout->digit_size == 4);
+
+    uint32_t base = (uint32_t)1 << layout->bits_per_digit;
+    size_t n, len = mpd_sizeinbase(x, base);
```

Comment:
```suggestion
    size_t len = mpd_sizeinbase(x, base);
```

Do you want maybe to throw a comment saying that the result won't be underestimated + a link to the GH comment you made?

Reply:
Comment like this right below, you edited.  Do we need something more?  Do we need move it to a different place?

I would prefer to keep `n` declaration here, it's not too far from actual usage.

Reply:
I think it's fine now. Linking a GH comment usually takes the entire line. Ok dor keeping `n` here as well.